pkg/resource/edit/operations.go

```diff
@@ -63,43 +63,65 @@ func DeleteResource(
 		return err
 	}
 
-	dg := graph.NewDependencyGraph(snapshot.Resources)
-	deleteSet := map[resource.URN]bool{
-		condemnedRes.URN: true,
+	var numSameURN int
+	for _, res := range snapshot.Resources {
+		if res.URN != condemnedRes.URN {
+			continue
+		}
+		numSameURN++
 	}
 
-	if dependencies := dg.DependingOn(condemnedRes, nil, true); len(dependencies) != 0 {
-		if !targetDependents {
-			return ResourceHasDependenciesError{Condemned: condemnedRes, Dependencies: dependencies}
-		}
-		for _, dep := range dependencies {
-			err := handleProtected(dep)
-			if err != nil {
-				return err
+	deleteSet := map[resource.URN]struct{}{}
+
+	isUniqueURN := numSameURN <= 1
+	// If there's only one resource (or less), determine dependencies to be deleted from state.
+	if isUniqueURN {
+		// condemnedRes.URN is unique. We can safely delete it by URN.
+		deleteSet[condemnedRes.URN] = struct{}{}
+		dg := graph.NewDependencyGraph(snapshot.Resources)
+
+		if deps := dg.DependingOn(condemnedRes, nil, true); len(deps) != 0 {
+			if !targetDependents {
+				return ResourceHasDependenciesError{Condemned: condemnedRes, Dependencies: deps}
```

Comment:
Out of scope for this PR but:

Error structs should almost always use pointer receivers for the Error method (which would make this `return &ResourceHasDependenciesError{..}`).
That keeps them comparable via references with `==`, even if they contain uncomparable fields like slices (which this one does).

Right now, `errors.Is(err, err)` on a ResourceHasDependenciesError will fail even if it's the same object.
(demo: https://go.dev/play/p/2OJl_QsL_cH)